File: sdk/typescript/examples/basic-usage.ts
/**
 * Basic usage example for the Facto SDK.
 *
 * Run with: npx ts-node examples/basic-usage.ts
 */

import { FactoClient, verifyEvent, toWireFormat } from '../src';

// Simulate an LLM call
async function simulateLLMCall(prompt: string): Promise<string> {
  await new Promise((resolve) => setTimeout(resolve, 100));
  return `Response to: ${prompt}`;
}

// Simulate a tool call
async function simulateToolCall(
  toolName: string,
  args: Record<string, unknown>
): Promise<Record<string, unknown>> {
  await new Promise((resolve) => setTimeout(resolve, 50));
  return { result: `Executed ${toolName}`, args };
}

async function main() {
  console.log('Initializing Facto client...');

  // Initialize the client
  const client = new FactoClient({
    endpoint: 'http://localhost:8080',
    agentId: 'example-agent-001',
    tags: { environment: 'development', version: '1.0.0' },
  });

  console.log(`Session ID: ${client.sessionId}`);

  // Method 1: Manual recording
  console.log('\n1. Recording events manually...');
  const initFactoId = await client.record({
    actionType: 'initialization',
    inputData: { config: { model: 'gpt-4', temperature: 0.7 } },
    outputData: { status: 'ready' },
    executionMeta: {
      modelId: 'gpt-4',
      temperature: 0.7,
    },
  });
  console.log(`   Recorded initialization event: ${initFactoId}`);

  // Method 2: Using factod wrapper
  console.log('\n2. Using factod wrapper...');
  const factodLLMCall = client.factod(
    'llm_call',
    async (prompt: string) => await simulateLLMCall(prompt)
  );
  const response = await factodLLMCall('What is the capital of France?');
  console.log(`   LLM response: ${response}`);

  // Method 3: Using startFacto for manual control
  console.log('\n3. Using startFacto context...');
  const ctx = client.startFacto('tool_use', {
    tool: 'web_search',
    query: 'latest AI news',
  });

  try {
    const toolResult = await simulateToolCall('web_search', {
      query: 'latest AI news',
    });
    ctx.setOutput(toolResult);
    ctx.setStatus('success');
  } catch (error) {
    if (error instanceof Error) {
      ctx.setError(error);
    }
  } finally {
    await ctx.end();
  }
  console.log(`   Recorded tool use event: ${ctx.factoId}`);

  // Method 4: Nested events with parent-child relationship
  console.log('\n4. Recording nested events...');
  const parentCtx = client.startFacto('agent_task', { task: 'research' });

  // Child event 1
  await client.record({
    actionType: 'llm_call',
    inputData: { prompt: 'Analyze the research topic' },
    outputData: { analysis: 'Topic is interesting' },
    parentFactoId: parentCtx.factoId,
  });

  // Child event 2
  await client.record({
    actionType: 'tool_use',
    inputData: { tool: 'database_query' },
    outputData: { records: 10 },
    parentFactoId: parentCtx.factoId,
  });

  parentCtx.setOutput({ status: 'completed', children: 2 });
  await parentCtx.end();
  console.log(`   Recorded parent event: ${parentCtx.factoId}`);

  // Method 5: Error handling
  console.log('\n5. Recording error events...');
  const errorCtx = client.startFacto('risky_operation', { action: 'divide' });
  try {
    throw new Error('Division by zero');
  } catch (error) {
    if (error instanceof Error) {
      errorCtx.setError(error);
    }
  }
  await errorCtx.end();
  console.log(`   Recorded error event: ${errorCtx.factoId}`);

  // Demonstrate event verification
  console.log('\n6. Creating and verifying an event...');
  const verifyFactoId = await client.record({
    actionType: 'verification_test',
    inputData: { test: true },
    outputData: { verified: true },
  });
  console.log(`   Created event for verification: ${verifyFactoId}`);

  // Flush events
  console.log('\n7. Flushing events...');
  try {
    await client.flush();
    console.log('   Events flushed successfully!');
  } catch (error) {
    console.log(`   Note: Could not connect to server (${error})`);
    console.log(
      '   Events are batched and would be sent when server is available'
    );
  }

  // Close client
  await client.close();
  console.log('\nClient closed.');
}

main().catch(console.error);

File: sdk/python/examples/basic_usage.py
#!/usr/bin/env python3
"""
Basic usage example for the Facto SDK.

This example demonstrates:
1. Initializing the client
2. Recording events manually
3. Using the context manager
4. Using the decorator
5. Verifying events
"""

import time
from facto_sdk import (
    FactoClient,
    FactoConfig,
    ExecutionMeta,
    verify_event,
)


def simulate_llm_call(prompt: str) -> str:
    """Simulate an LLM call."""
    time.sleep(0.1)  # Simulate latency
    return f"Response to: {prompt}"


def simulate_tool_call(tool_name: str, args: dict) -> dict:
    """Simulate a tool call."""
    time.sleep(0.05)
    return {"result": f"Executed {tool_name}", "args": args}


def main():
    # Initialize the client
    print("Initializing Facto client...")
    config = FactoConfig(
        endpoint="http://127.0.0.1:8080",
        agent_id="example-agent-001",
        tags={"environment": "development", "version": "1.0.0"},
    )
    client = FactoClient(config)
    print(f"Session ID: {config.session_id}")

    # Method 1: Manual recording
    print("\n1. Recording events manually...")
    facto_id = client.record(
        action_type="initialization",
        input_data={"config": {"model": "gpt-4", "temperature": 0.7}},
        output_data={"status": "ready"},
        execution_meta=ExecutionMeta(
            model_id="gpt-4",
            temperature=0.7,
        ),
    )
    print(f"   Recorded initialization event: {facto_id}")

    # Method 2: Context manager
    print("\n2. Using context manager...")
    with client.facto(
        "llm_call",
        input_data={"prompt": "What is the capital of France?"},
        execution_meta=ExecutionMeta(model_id="gpt-4", temperature=0.7),
    ) as ctx:
        response = simulate_llm_call("What is the capital of France?")
        ctx.output = {"response": response, "tokens": 42}
    print(f"   Recorded LLM call event: {ctx.facto_id}")

    # Method 3: Decorator
    print("\n3. Using decorator...")

    @client.factod("tool_use")
    def search_web(query: str) -> dict:
        """Search the web for a query."""
        return simulate_tool_call("web_search", {"query": query})

    result = search_web("latest AI news")
    print(f"   Recorded tool use event for web search")

    # Multiple events with parent-child relationship
    print("\n4. Recording nested events...")
    with client.facto("agent_task", input_data={"task": "research"}) as parent_ctx:
        # Child event 1
        with client.facto(
            "llm_call",
            input_data={"prompt": "Analyze the research topic"},
            parent_facto_id=parent_ctx.facto_id,
        ) as child1:
            child1.output = {"analysis": "Topic is interesting"}

        # Child event 2
        with client.facto(
            "tool_use",
            input_data={"tool": "database_query"},
            parent_facto_id=parent_ctx.facto_id,
        ) as child2:
            child2.output = {"records": 10}

        parent_ctx.output = {"status": "completed", "children": 2}
    print(f"   Recorded parent event: {parent_ctx.facto_id}")

    # Demonstrate error handling
    print("\n5. Recording error events...")
    try:
        with client.facto("risky_operation", input_data={"action": "divide"}) as ctx:
            raise ValueError("Division by zero")
    except ValueError:
        pass  # Error is automatically recorded
    print(f"   Recorded error event: {ctx.facto_id}")

    # Verify an event
    print("\n6. Verifying event integrity...")
    # Get the last event from the batch
    if client._batch:
        last_event = client._batch[-1]
        event_dict = last_event.to_dict()
        hash_valid, sig_valid = verify_event(event_dict)
        print(f"   Hash valid: {hash_valid}")
        print(f"   Signature valid: {sig_valid}")

    # Flush and close
    print("\n7. Flushing events...")
    try:
        client.flush()
        print("   Events flushed successfully!")
    except Exception as e:
        print(f"   Note: Could not connect to server ({e})")
        print("   Events are still in the batch and would be sent when server is available")

    # Print batch summary
    print(f"\nTotal events recorded: {len(client._batch)}")
    for event in client._batch:
        print(f"  - {event.action_type}: {event.facto_id}")

    client.close()
    print("\nClient closed.")


if __name__ == "__main__":
    main()
